cli/inspector.rs

```diff
@@ -0,0 +1,204 @@
+            client.on_connection()
+          }))
+        }));
+
+    let json = warp::path("json").map(|| warp::reply::json(&*RESPONSE_JSON));
+
+    let version =
+      path!("json" / "version").map(|| warp::reply::json(&*RESPONSE_VERSION));
+
+    let routes = websocket.or(version).or(json);
+
+    let endpoint = format!("{}:{}", HOST, PORT);
+    let addr = endpoint.parse::<SocketAddrV4>().unwrap();
+
+    warp::serve(routes).bind(addr)
+  }
+
+  pub fn start(&mut self) {
+    let mut started = self.started.lock().unwrap();
+
+    if *started || !self.enable {
+      return;
+    }
+
+    *started = true;
+
+    self.server_spawn_handle = Some(spawn(
+      self.serve(),
+      &tokio_executor::DefaultExecutor::current(),
+    ));
+
+    println!("Debugger listening on ws://{}:{}/{}", HOST, PORT, UUID);
+
+    if self.pause_on_start {
+      self
+        .ready_rx
+        .lock()
+        .unwrap()
+        .recv()
+        .expect("Error waiting for inspector server to start.");
+      // TODO(mtharrison): This is to allow v8 to ingest some inspector messages - find a more reliable way
+      std::thread::sleep(std::time::Duration::from_secs(1));
+      println!("Inspector frontend connected.");
+    }
+  }
+}
+
+impl Drop for Inspector {
+  fn drop(&mut self) {
+    if *self.connected.lock().unwrap() == true {
+      println!("Waiting for debugger to disconnect...");
+    }
+  }
+}
+
+#[derive(Clone)]
+pub struct ClientState {
+  sender: Arc<Mutex<Sender<String>>>,
+  receiver: Arc<Mutex<Receiver<String>>>,
+  ready_tx: Arc<Mutex<Sender<()>>>,
+  connected: Arc<Mutex<bool>>,
+}
+
+pub struct Client {
+  state: ClientState,
+  socket: WebSocket,
+}
+
+impl Client {
+
+  fn new(state: ClientState, socket: WebSocket) -> Client {
+    Client {
+      state, socket
+    }
+  }
+
+  fn on_connection(self) -> impl Future<Item = (), Error = ()> {
+
+    let socket = self.socket;
+    let sender = self.state.sender;
+    let receiver = self.state.receiver;
+    let connected = self.state.connected;
+    let ready_tx = self.state.ready_tx;
+
+    let (mut user_ws_tx, user_ws_rx) = socket.split();
+
+    let fut_rx = user_ws_rx
+      .for_each(move |msg| {
+        let msg_str = msg.to_str().unwrap();
+        sender
+          .lock()
+          .unwrap()
+          .send(msg_str.to_owned())
+          .unwrap_or_else(|err| println!("Err: {}", err));
+        Ok(())
+      }).map_err(|_| {});
+
+    // TODO(mtharrison): This is a mess. There must be a better way to do this - maybe use async channels or wrap them with a stream?
+
+    std::thread::spawn(move || {
+      let receiver = receiver.lock().unwrap();
+      loop {
+        let received = receiver.recv();
+        if let Ok(msg) = received {
+          let _ = ready_tx.lock().unwrap().send(());
+          *connected.lock().unwrap() = true;
+          user_ws_tx = user_ws_tx.send(Message::text(msg)).wait().unwrap();
+        }
+      }
+    });
```

Comment:
> TODO(mtharrison): This is a mess. There must be a better way to do this - maybe use async channels or wrap them with a stream?

Actually I think starting a thread and synchronously receiving is appropriate in this case.

Reply:
I am not sure that I understand rust enough to figure out what is happening in this code.

Inspector assumes that all V8Inspector functions are called on the main thread.
In case if V8 is running some busy loop right now, there is a V8 API to request interrupt from any thread: https://cs.chromium.org/chromium/src/v8/include/v8.h?q=f:v8.h&sq=package:chromium&g=0&l=8222

I think that in Node on the new message we try to post a message to the main thread and RequestInterrupt at the same time, first who succeed will process a message.

Reply:
Relevant Node code: https://github.com/nodejs/node/blob/2abdfc5d7eba620a84a1aea3804beb6de28e19fa/src/inspector/main_thread_interface.cc#L227

We send RequestInterrupt in case Node is running continuous code block (e.g. `while(true){}`) and we trigger libuv in case Node.js is waiting for the IO event (e.g. web service waiting for a connection).